### PR TITLE
add dict repository and repo test cases

### PR DESCRIPTION
**Pull request by abhishek-ram**
PR for issue #25 - Dict Repository Impl

### CODE REVIEW

File: src/protean/core/entity.py [hunk 1]
```diff
@@ -71,9 +71,10 @@ def __init__(self, *template, **kwargs):
         """
 
         self.errors = {}
-        fields = self.get_fields()
+        self.declared_fields = self.get_fields()
```
Comment (subhashb) [inline src/protean/core/entity.py]: Like the distinction! 👍

File: tests/support/dict_repo.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+from operator import itemgetter
+
+from protean.core.entity import Entity
+from protean.core.exceptions import DuplicateObjectError
+from protean.core.repository import RepositoryFactory, Repository, \
+    RepositorySchema, Pagination
+
+
+class DictRepository(Repository):
```
Comment (subhashb) [inline tests/support/dict_repo.py]: Awesome! 👍

File: tests/support/dict_repo.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            del_count += 1
+        except KeyError:
+            pass
+        return del_count
+
+
+class DictSchema(RepositorySchema):
+
+    class Meta:
+        entity = Entity
```
Comment (subhashb) [inline tests/support/dict_repo.py]: Shouldn't this be `entity = Dog`

File: tests/support/dict_repo.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+class DictRepositoryFactory(RepositoryFactory):
+
+    def get_connection(self):
+        """ Return the dictionary database object """
+        if not self._connection:
+            self._connection = defaultdict(dict)
+        return self._connection
+
+
+drf = DictRepositoryFactory()
```
Comment (subhashb) [inline tests/support/dict_repo.py]: Eventually, `DictRepositoryFactory` should be a singleton that accepts an entity class and supplies the prepared model/schema/repo (already connected to the database). Correct?

Comment (abhishek-ram) [reply tests/support/dict_repo.py]: Yes it is meant to be a singleton for registering a schema with a repository and can return the repository for a schema

File: tests/support/dict_repo.py (hunk 3 above)
Comment (abhishek-ram) [reply tests/support/dict_repo.py]: I have resolved this now

**Pull request merged**